src/provider/model/Model.ts

```diff
@@ -277,17 +244,48 @@ export class Model {
         return this.windows.filter(predicate);
     }
 
-    private getOrCreateExpectedWindow(identity: Identity): ExpectedWindow {
+    private getOrCreateExpectedWindow(identity: Identity, windowSeen?: 'seen'): ExpectedWindow {
         const id = getId(identity);
 
         if (this._expectedWindows.has(id)) {
             return this._expectedWindows.get(id)!;
         } else {
-            const expectedWindow: ExpectedWindow = {
-                pendingDeferredPromise: new DeferredPromise<Boxed<Promise<AppWindow>>>(),
-                createdDeferredPromise: new DeferredPromise<void>(),
-                registerDeferredPromise: new DeferredPromise<AppWindow>()
-            };
+            // Create a promise that resolves when the window has registered, or rejects if it is closed
+            const registered = Promise.race([
+                untilTrue((args) => {
+                    return !!this._windowsById[id];
+                }, this._onWindowRegisteredInternal).then(() => {
+                    return this._windowsById[id];
+                }),
+                untilTrue((args) => {
+                    return !!args && getId(args[0]) === id;
+                }, this._environment.windowClosed).then(() => {
+                    throw new Error(EXPECT_CLOSED_MESSAGE);
+                })
+            ]);
+
+            // Create a promise that resolves when the window has connected, or rejects if it is closed
+            const connected = Promise.race([
+                untilTrue((args) => {
+                    return !!args && getId(args[0]) === id;
+                }, this._apiHandler.onConnection),
+                untilTrue((args) => {
+                    return !!args && getId(args[0]) === id;
```

Comment:
Could you change the predicate signature in `untilTrue` to be:
```ts
predicate: (...args: A) => boolean
```
And then here we can explicitly list the args by name instead of using array indexing, makes it easier to read:
```ts
untilTrue((value: string) => {
	return !!value && getId(value) === id;
...
})
```

Reply:
Would prefer to leave as is - this means in all cases the predicate function can tell whether it's being called in response to the signal or not, and we're not relying on the signal payload not taking on certain values